tilings/assumptions.py

```diff
@@ -72,3 +80,119 @@ def __str__(self):
+
+
+class ComponentAssumption(TrackingAssumption):
+    """
+    An assumption used to keep track of the number of omponents in a
+    region of a tiling.
+
+    In order to inherit from TrackingAssumption, the set of cells should be
+    given as a set of length 1 gridded perms using each cell. This ensures
+    most strategies work without change.
+    """
+
+    def __init__(self, gps: Iterable[GriddedPerm]):
+        super().__init__(gps)
+        assert all(len(gp) == 1 for gp in self.gps)
+        self.cells = frozenset(gp.pos[0] for gp in self.gps)
+
+    @abc.abstractproperty
+    def decomposition(self):
+        pass
+
+    def get_value(self, gp: GriddedPerm) -> int:
+        """
+        Return the number of occurrences of each of the gridded perms being track in
+        the gridded perm gp.
+        """
+        subgp = gp.get_gridded_perm_in_cells(self.cells)
+        return len(self.decomposition(subgp.patt))
+
+    def to_jsonable(self) -> dict:
+        return {"cells": tuple(self.cells)}
+
+    @classmethod
+    def from_dict(
+        cls, d: dict
+    ) -> Union["SumComponentAssumption", "SkewComponentAssumption"]:
+        gps = [GriddedPerm.single_cell(Perm((0,)), cell) for cell in d["cells"]]
+        if d.pop("sum"):
+            return SumComponentAssumption(gps)
+        return SkewComponentAssumption(gps)
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, ComponentAssumption) and self.__class__ == other.__class__:
+            return bool(self.gps == other.gps)
+        return NotImplemented
+
+    def __repr__(self) -> str:
+        return self.__class__.__name__ + "({})".format(self.gps)
+
+    def __str__(self):
+        return f"can count components in cells {self.cells}"
+
+    def __hash__(self) -> int:
+        return hash(self.gps)
+
+
+class SumComponentAssumption(ComponentAssumption):
+    @staticmethod
+    def sum_decomposition(perm: Perm) -> List[Perm]:
+        """
+        Return the sum decomposition of the permutation.
+        TODO: this method should be on permuta.
+        """
+        res: List[Perm] = []
+        max_val = -1
+        curr_block_start_idx = 0
+        for idx, val in enumerate(perm):
+            max_val = max(max_val, val)
+            if idx == max_val:
+                res.append(Perm.to_standard(perm[curr_block_start_idx : idx + 1]))
+                curr_block_start_idx = idx + 1
+        return res
+
+    decomposition = sum_decomposition
+
+    def to_jsonable(self) -> dict:
+        d = super().to_jsonable()
+        d["sum"] = True
+        return d
+
+    def __str__(self):
+        return f"can count sum components in cells {self.cells}"
+
+    def __hash__(self) -> int:
+        return hash(self.gps)
+
+
+class SkewComponentAssumption(ComponentAssumption):
+    @staticmethod
+    def skew_decomposition(perm: Perm) -> List[Perm]:
+        """
+        Return the skew decomposition of the permutation.
+        TODO: this method should be on permuta.
```

Comment:
Did it https://github.com/PermutaTriangle/Permuta/pull/114

Reply:
I changed to use this, but of course now it won't be able to pass travis tests as that points to master I think.

Reply:
Yep we should release a new version of permuta

Reply:
You should also remove the two static methods

Reply:
Thanks!

Reply:
You forgot to remove the static methods :)

Reply:
good spot!